scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +171,373 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
+        The intent is for the interface to be natural, but the detailed rules
+        to achieve this are complicated.
+
+        - If `qmc_engine` is ``None``, a `scipy.stats.qmc.Halton` instance is
+          created with dimension `d`. If `d` is not provided, ``d=1``.
+        - If `qmc_engine` is not ``None`` and `d` is ``None``, `d` is
+          determined from the dimension of the `qmc_engine`.
+        - If `qmc_engine` is not ``None`` and `d` is not ``None`` but the
+          dimensions are inconsistent, a ``ValueError`` is raised.
+        - If `size` is ``None`` and `d` is 1 (after the logic above), a scalar
+          is returned.
+        - If `size` is not ``None`` and `d` is 1 (after the logic above),
+          an array of shape `size` (or ``(size,)``) is returned.
+        - If `size` is not ``None`` and `d` is greater than 1 (after the logic
+          above), an array of shape ``size + (d,)`` (or ``(size, d)``)
+          is returned.
+
+        The elements of the returned array are part of a low-discrepancy
+        sequence. If `d` is 1, this means that none of the samples are truly
+        independent. If `d` > 1, each slice ``rvs[..., i]`` will be of a
+        quasi-independent sequence; see `scipy.stats.qmc.QMCEngine` for
+        details. Note that when `d` > 1, the samples returned are still those
+        of the provided univariate distribution, not a multivariate
+        generalization of that distribution.
+
+        """
+        # Input validation for `qmc_engine` and `d`
+        # Error messages for invalid `d` are raised by QMCEngine
+        # we could probably use a stats.qmc.check_qrandom_state
+        if isinstance(qmc_engine, stats.qmc.QMCEngine):
+            message = "`d` must be consistent with dimension of `qmc_engine`."
+            if d is not None and qmc_engine.d != d:
+                raise ValueError(message)
+            d = qmc_engine.d if d is None else d
+        elif qmc_engine is None:
+            d = 1 if d is None else d
+            qmc_engine = stats.qmc.Halton(d)
+        else:
+            message = ("`qmc_engine` must be an instance of "
+                       "`scipy.stats.qmc.QMCEngine` or `None`.")
+            raise ValueError(message)
+
+        # `rvs` is flexible about whether `size` is an int or tuple, so this
+        # should be, too.
+        try:
+            tuple_size = tuple(size)
+        except TypeError:
+            tuple_size = (size,)
+
+        # Get uniform QRVS from qmc_random and transform it
+        uniform = qmc_engine.random(np.prod(tuple_size) or 1)
+        qrvs = self.ppf(uniform)
+
+        # Output reshaping for user convenience
+        if size is None:
+            return qrvs.squeeze()[()]
+        else:
+            if d == 1:
+                return qrvs.reshape(tuple_size)
+            else:
+                return qrvs.reshape(tuple_size + (d,))
+
+
+def _fni_input_validation(dist, tol, max_intervals):
```

Comment:
Since we check the instance type of the `QMCEngine`, we might want to check here that `dist` is of correct type, or just that it has `ppf, isf, cdf`.

Reply:
Yes, i was thinking about that yesterday. I can do that.